framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/jobs/JobManager.java

```diff
@@ -26,34 +38,62 @@ public void removeJobListener(IJobListener l) {
 	}
 
 	@Override
-	public void addJob(IJob job) {
+	public IJob scheduleJob(String jobName, IRunnableWithProgress runnable) {
+		SimpleJob job = new SimpleJob(jobName, generateJobId(), runnable);
 		currentJobs.put(job.getId(), job);
-		ArrayList<IJobListener> tmp = new ArrayList<>(listeners);
-		for( IJobListener l : tmp ) {
-			l.jobAdded(job);
-		}
+		fireJobAdded(job);
+		schedule(job);
+		return job;
 	}
 
 	@Override
-	public void removeJob(IJob job, IStatus status) {
-		currentJobs.remove(job.getId());
+	public IJob scheduleJob(String jobName, IStatusRunnableWithProgress runnable) {
+		SimpleJob job = new SimpleJob(jobName, generateJobId(), runnable);
+		currentJobs.put(job.getId(), job);
+		fireJobAdded(job);
+		schedule(job);
+		return job;
+	}
+	
+	private void fireJobAdded(IJob job) {
 		ArrayList<IJobListener> tmp = new ArrayList<>(listeners);
 		for( IJobListener l : tmp ) {
-			l.jobRemoved(job, status);
+			l.jobAdded(job);
 		}
 	}
 
+	private void schedule(SimpleJob job) {
+		executor.execute(() -> {
+			IStatus s = null;
+			try {
+				s = job.run();
+			} catch(Exception e) {
+				s = new Status(IStatus.ERROR, ServerCoreActivator.BUNDLE_ID, e.getMessage(), e);
+			}
+			jobComplete(job, s);
+		});
+	}
+	
 	@Override
-	public void progressChanged(IJob job, double work) {
+	public void cancel(IJob job) {
+		job.getProgressMonitor().setCanceled(true);
```

Comment:
are we making sure that there always is a progress monitor so that there's no chance for an NPE here?